src/relayclient/KnownRelaysManager.ts

```diff
@@ -103,14 +104,55 @@ export default class KnownRelaysManager implements IKnownRelaysManager {
     return origRelays.filter(this.relayFilter)
   }
 
+  splitRange (fromBlock: number, toBlock: number, splits: number): Array<{fromBlock: number, toBlock: number}> {
+    const totalBlocks = toBlock - fromBlock + 1
+    const splitSize = Math.ceil(totalBlocks / splits)
+
+    const ret: Array<{ fromBlock: number, toBlock: number }> = []
+    let b
+    for (b = fromBlock; b < toBlock; b += splitSize) {
+      ret.push({ fromBlock: b, toBlock: Math.min(toBlock, b + splitSize - 1) })
+    }
+    return ret
+  }
+
+  // return events from hub. split requested range into "window parts", to avoid
+  // fetching too many events at once.
+  async getPastEventsForHub (fromBlock: number, toBlock: number): Promise<EventData[]> {
+    let relayEventParts: any[]
+    while (true) {
+      const rangeParts = this.splitRange(fromBlock, toBlock, this.config.relayLookupWindowParts)
+      try {
+        // eslint-disable-next-line @typescript-eslint/promise-function-async
+        const getPastEventsPromises = rangeParts.map(({ fromBlock, toBlock }): Promise<any> =>
+          this.contractInteractor.getPastEventsForHub([], {
+            fromBlock,
+            toBlock
+          }))
+        relayEventParts = await Promise.all(getPastEventsPromises)
+        break
+      } catch (e) {
+        if (e.toString().match(/query returned more than/) != null &&
```

Comment:
Isn't this 100% specific to Infura? Is it ok to have Infura-specific error messages hard-coded in our code?

Reply:
Yes, it is... alas, I don't have any other way to recognize it.
(it has `"code":-32005`, but even that is not unique for this error)
AFAIK, the native node limits of geth/parity are much higher
(BTW: EtherScan API , which we currently don't support since its not exactly  json-rpc, return the FIRST 10000 responses, effectively paginate the response. I like that solution much better - even though its not 100% , but we will add it after moving to `ethers`, which supports etherscan API natively.)